### PR TITLE
Have time steppers overwrite in update_u again

### LINKED ISSUE #4399: Dense output with local time stepping gives incorrect results
# Bug reports:

### Expected behavior:

Dense output with local time stepping (LTS) does not have large errors that grow linearly with the time step size.

### Current behavior:

I have a tensor that is zero at t=0. Observing with local time stepping and dense output incorrectly gives an L2 norm of O(1e-6) at t=0.

Path to evolution: `ocean:/home/geoffrey/BBH/spectre/q1spin0_develop/111622/spectre_run/h1p9csph/enforce3con/nomemmonitor/densepossibelfail/ok4`
The run sets the 3-index constraint to zero at t=0, but the observed L2 norm in the reductions h5 file is 7e-6.

### Environment:

```
SpECTRE Version: 2022.11.15
Git description: v2022.04.04-397-g2a2ba3fae3
Git branch: Enforce3ConInNumericInitialData
Git hash: 2a2ba3fae3a13179e1bcca8fcf40bab328cd3ba8
Hostname: ocean.fullerton.edu
Host system: Linux-3.10.0-957.1.3.el7.x86_64
Host system version: 3.10.0-957.1.3.el7.x86_64
Host system processor: x86_64
Boost version: 1.79.0
Charm++ version: 7.0.0
CHARM_COMPILER: /opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/bin/charmc
CHARM_SHARED_LIBS: OFF
CHARM_BUILDING_BLOCKS:
CHARM_CC='clang'
CHARM_CXX='clang++'
CHARM_LD='clang'
CHARM_LDXX='clang++'
CHARM_CC_FLAGS=' -DCMK_GFORTRAN -D_REENTRANT -I/opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/include -D__CHARMC__=1 -U_FORTIFY_SOURCE -fno-stack-protector -ftls-model=initial-exec '
CHARM_CXX_FLAGS=' -DCMK_GFORTRAN -D_REENTRANT -I/opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/include -D__CHARMC__=1 -U_FORTIFY_SOURCE -fno-stack-protector -ftls-model=initial-exec -std=gnu++11 '
CHARM_LD_FLAGS=' -D_REENTRANT -ftls-model=initial-exec -rdynamic -L/opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/lib -lmoduleEveryLB -lckmetis -lck -lmemory-default -lthreads-default -lconverse -lm -lmemory-default -lthreads-default -lldb-rand -lconverse -lpthread -lckqt -libverbs -lz -ldl -lmoduleEveryLB -lmoduleNDMeshStreamer -lmodulecompletion -lz -lm /opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/lib/conv-static.o'
CHARM_LDXX_FLAGS=' -D_REENTRANT -ftls-model=initial-exec -rdynamic -L/opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/lib -lmoduleEveryLB -lckmetis -lck -lmemory-default -lthreads-default -lconverse -lm -lmemory-default -lthreads-default -lldb-rand -lconverse -lpthread -lckqt -libverbs -lz -ldl -lmoduleEveryLB -lmoduleNDMeshStreamer -lmodulecompletion -lz -lm /opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/lib/conv-static.o'
CHARMBIN='/opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/bin'
CHARMINC='/opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/include'
CHARMLIB='/opt/ohpc/pub/apps/charm_7.0.0_gnu11_clang_smp/verbs-linux-x86_64-smp-clang-gfortran/lib'
CHARMLIBSO=''
BLAS_LIBRARIES: /opt/ohpc/pub/apps/spack2022/opt/spack/linux-centos7-broadwell/gcc-11.3.0/openblas-0.3.20-tc4qxfvqf367fkbssorninewy27t4m76/lib/libopenblas.so
LAPACK_LIBRARIES: /opt/ohpc/pub/apps/spack2022/opt/spack/linux-centos7-broadwell/gcc-11.3.0/openblas-0.3.20-tc4qxfvqf367fkbssorninewy27t4m76/lib/libopenblas.so;-lpthread;-lm;-ldl
LIBXSMM version: 1.16.1
Blaze version: 3.8
GSL version: 2.7.1
HDF5 version: 1.12.2
jemalloc version: 5.2.1
Catch version: 2.13.8
NumPy version: 1.21.1
CMake version: 3.24.1
CMAKE_CXX_FLAGS: 
CMAKE_CXX_LINK_FLAGS:  -fuse-ld=lld
CMAKE_CXX_FLAGS_DEBUG: -g -DSPECTRE_DEBUG
CMAKE_CXX_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_C_FLAGS: 
CMAKE_C_FLAGS_DEBUG: -g
CMAKE_C_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_Fortran_FLAGS: 
CMAKE_Fortran_FLAGS_DEBUG: -g
CMAKE_Fortran_FLAGS_RELEASE: -O3 -DNDEBUG -O3
CMAKE_EXE_LINKER_FLAGS: 
CMAKE_BUILD_TYPE: Release
CMAKE_CXX_COMPILER: /opt/ohpc/pub/compiler/llvm/13.0.1/bin/clang++
CMAKE_CXX_COMPILER_VERSION: 13.0.1
CMAKE_C_COMPILER: /opt/ohpc/pub/compiler/llvm/13.0.1/bin/clang
CMAKE_C_COMPILER_VERSION: 13.0.1
CMAKE_Fortran_COMPILER: /opt/ohpc/pub/compiler/gcc/11.3.0/bin/gfortran
CMAKE_Fortran_COMPILER_VERSION: 11.3.0
Python version: 3.9.5
```

# Feature request:

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Desired feature:

- [ ] Detail 1
- [ ] Detail 2
- [ ] Detail 3

### PR DESCRIPTION
**Pull request by wthrowe**
It turns out that it is sometimes difficult for the caller to determine the correct value to pass into the function.

This largely reverts bf4fd0a409c76996a104cb8750b4b9fb8e190c58.

Fixes #4399 

@geoffrey4444 This should remove the need for the dense output hack.  Can you test this and verify that your results look good?

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Review by nilsdeppe — DISMISSED** [review]

**Comment by geoffrey4444** [comment]
Sorry, but it looks to me like this PR does not fix the dense output bug #4399. I tried running a BBH with the 3-index constraint set to zero at t=0. I used [this branch](https://github.com/sxs-collaboration/spectre/compare/develop...geoffrey4444:spectre:BBH_020823_bugfix) at 3 different revisions: , and one of the following:

1: nothing (no additional commit) — revision 9200dd5633
2: the commit in this PR — revision 291a228ae9
3: the hack commit you previously suggested to fix the bug — revision b9fbf8e178

I verified that spectre.out reported the correct revision in each case. At t=0, here's the L2 norm of the 3-index constraint for each case:

1.  8.81945e-06
2. 8.81945e-06
3. 1.60104e-17

**Comment by wthrowe** [comment]
OK, try now.  An old optimization was bypassing the fix when run at t=0.

This fix is not really related to the update_u changes, so I'm going to leave it as a separate commit.

File: src/Time/TimeSteppers/AdamsBashforth.cpp [hunk 1]
```diff
@@ -366,6 +366,9 @@ void AdamsBashforth::boundary_dense_output_impl(
     const gsl::not_null<T*> result,
     const TimeSteppers::BoundaryHistoryEvaluator<T>& coupling,
     const double time) const {
+  if ((coupling.local_end() - 1)->value() == time) {
+    return;
```
Comment (knelli2) [inline src/Time/TimeSteppers/AdamsBashforth.cpp]: I think this means if we are at the most recent history time, just return (i.e. don't do dense output), but I'm not 100% certain. Could you add a comment?

**Review by knelli2 — CHANGES_REQUESTED** [review]

File: src/Time/TimeSteppers/AdamsBashforth.cpp (hunk 1 above)
Comment (wthrowe) [reply src/Time/TimeSteppers/AdamsBashforth.cpp]: You are correct.  I've added a comment.

**Review by knelli2 — APPROVED** [review]
I'm happy with the changes, but @geoffrey4444 should confirm again that these changes now fix the LTS dense output bug

**Comment by geoffrey4444** [comment]
> OK, try now. An old optimization was bypassing the fix when run at t=0.
> 
> This fix is not really related to the update_u changes, so I'm going to leave it as a separate commit.

I will try this again. Thanks!

**Review by geoffrey4444 — APPROVED** [review]
I did not review the code line by line but did try this again, and now the dense and not-dense output for the 3-index constraints agree, even at t=0. 

image


Without this fix, the constraints are not zeroed correctly at t=0, and they level off to a larger value by time t=10 (though strangely even without the fix, I don't see the linear growth I saw back in November when I first spotted the symptoms of the bug this PR fixes).

I see small differences vs the hack fix, but I don't think they are significant.

So I'm happy to approve.

**Review by nilsdeppe — APPROVED** [review]

**Comment by wthrowe** [comment]
If you're doing filtering, I'm guessing the hack and this version interact with that differently.  If not, I don't immediately see a reason they would differ, but I wouldn't worry about it for the moment.

**Pull request merged**